awx/main/migrations/0124_event_partitions.py

```diff
@@ -0,0 +1,133 @@
+from datetime import datetime
+
+from django.db import migrations, models, connection
+from django.utils.timezone import now
+
+from awx.main.utils.common import create_partition
+
+
+def migrate_event_data(apps, schema_editor):
+    # see: https://github.com/ansible/awx/issues/9039
+    #
+    # the goal of this function is -- for each job event table -- to:
+    # - create a parent partition table
+    # - .. with a single partition
+    # - .. that includes all existing job events
+    #
+    # the new main_jobevent_parent table should have a new
+    # denormalized column, job_created, this is used as a
+    # basis for partitioning job event rows
+    #
+    # The initial partion will be a unique case. After
+    # the migration is completed, awx should create
+    # new partitions on an hourly basis, as needed.
+    # All events for a given job should be placed in
+    # a partition based on the job's _created time_.
+
+    for tblname in (
+        'main_jobevent', 'main_inventoryupdateevent',
+        'main_projectupdateevent', 'main_adhoccommandevent',
+        'main_systemjobevent'
+    ):
+        with connection.cursor() as cursor:
+            # mark existing table as *_old;
+            # we will drop this table after its data
+            # has been moved over
+            cursor.execute(
+                f'ALTER TABLE {tblname} RENAME TO {tblname}_old'
+            )
+
+            # drop primary key constraint; in a partioned table
+            # constraints must include the partition key itself
+            # TODO: do more generic search for pkey constraints
+            # instead of hardcoding this one that applies to main_jobevent
+            cursor.execute(
+                f'ALTER TABLE {tblname}_old DROP CONSTRAINT {tblname}_pkey1'
```

Comment:
Reading pg docs, this is a little unclear to me.  Does the *primary key* have to include the partition column, or just **any** indexed constraint?  There's almost certainly going to be situations where we want to search for a specific event based on ID only e.g., `/api/v2/jobs/N/job_events/Y/` and it's going to be super slow if this column isn't the key.

e.g.,

Does this still work if we **keep** `id` as the primary key, and *instead* do:

```sql
ALTER TABLE {tblname} ADD CONSTRAINT {tblname}_unique_id_job_created UNIQUE (id, job_created);
```

If this works w/ partitions, I'd prefer it.

Reply:
> Does this still work if we keep id as the primary key

No, this is not an option; postgres refuses to let you claim a field as being unique _unless_ you use a composite primary key and include the partition field (job_created)

Reply:
okay, found my comment where I posted my findings on this:
https://github.com/ansible/awx/issues/9259#issuecomment-776325677

Reply:
What I _think_ pg will try to do using the tables exactly the way they are today is:
* run a parallel search of each of the partitions associated with the parent job event table
* for each partition, use _that table's_ index for id / job_created

I tried running a query with `explain analyze` just now, but I hardly have any events in my db. It's showing a sequential scan, but my sense is that's probably because the postgres planner knew we hardly had any events and didn't bother with the index.

cc @Ladas for fact-checking because this is a pretty wild guess on my part

Reply:
@ryanpetrello @jladdjr so for query like `/api/v2/jobs/N/job_events/Y/` you will still need to use the partition key `job_created`, otherwise PG will scan all partitions and that will always be slow.

So the rule of thumb is using the partitioning key in all queries, when using partitions. (Well there are e.g. cases where you'd want to scan all the data in parallel, but postgres does pretty good parallel seq scan even without partitions.)

For this specific usecase PG requires unique constraint/primary key to include the partitioning key. And it should use the index on id with `(id, job_created)`

-------

@jladdjr yeah, the index/seq scan is dependent on what cost will PG compute. Which in simplified terms takes into account what percentage of the table PG thinks you'll be looking at, based on statistics it keeps. (very simply said, index should be used when querying ~5% of the table). 

Not sure if you're using default costs for PG. The most important is usually `random_page_cost`, which is by default 4.0, which is good for magnetic disks (index access is expensive). I set it 1.1 for ssd. Do you recommend PG config to users in docs? Or this is up to users?